### PR TITLE
Parquet-313: Implement 3 level list writing rule for Parquet-Thrift

### PR DESCRIPTION
**Pull request by SinghAsDev**
Parquet-thrift does not write three level lists, unlike other object models in the project. This JIRA should add 3 level list writing rule's implementation for parquet-thrift.

### CODE REVIEW

**Comment by rdblue** [comment]
@SinghAsDev can you look into the test failures and get this branch passing?

**Comment by SinghAsDev** [comment]
@rdblue this should be fixed now.

**Comment by SinghAsDev** [comment]
The build with hadoop-2 profile failed. Not sure why. I tried locally and it passes.

**Comment by rdblue** [comment]
@SinghAsDev, the error messages from the Travis logs are:

```
Tests in error: 
  testThriftOptionalFieldsWithReadProjectionUsingParquetSchema(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testReorderdOptionalFields(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testProjectOutOptionalFields(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredMaps(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValuePrimitive(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueStruct(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueNestedPrim(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testDropMapValueNestedStruct(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredLists(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInRequiredSets(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullInPrimitiveValues(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testPullingInRequiredStructWithFilter(org.apache.parquet.hadoop.thrift.TestParquetToThriftReadWriteAndProjection): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFile(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteStatistics(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileListOfMap(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileMapOfList(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileMapOfLists(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileWithThreeLevelsList(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
  testWriteFileListOfMapWithThreeLevelLists(org.apache.parquet.hadoop.thrift.TestThriftToParquetFileWriter): Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
```

`TaskAttemptContext` was changed from a class to an interface in Hadoop 2. That's a binary incompatible change, so we have to use shims, in `ContextUtil`, to dynamically resolve the right method. You're probably getting the configuration from the context somewhere and need to use the helper from `ContextUtil`.

**Comment by SinghAsDev** [comment]
Hey @rdblue would it be possible for you to review this?

File: parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java
Comment (piyushnarang) [inline parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java]: maybe pass an empty config rather than a null object?

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java]: empty config instead of null?

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java]: here too avoid null

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java]: can drop the `this.` for this line and the next right?

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java]: here too can drop the `this` for parquetWriteProtoCol and configuration.

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java]: empty instead of null?

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java]: nit: add a space after `Protocol` and before `{`

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java]: nit: add a space after `Protocol` and before `{`

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java]: we could refactor this out into a method `createListWriteProtocol(writeLevel3, columnIO, type, returnClause)` that way we don't have this largely identical repeated block

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java
Comment (piyushnarang) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: avoid null?

**Comment by SinghAsDev** [comment]
@piyushnarang your feedbacks should be addressed with latest change.

**Comment by SinghAsDev** [comment]
@rdblue / @isnotinvain I would highly appreciate if you guys can review this and help in merging this. This will avoid us from creating an internal fork.

File: parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java
Comment (nandorKollar) [inline parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java]: thriftStruct is an unused variable, the result of this conversion is not used, could you please delete this?

Comment (SinghAsDev) [reply parquet-scrooge/src/main/java/org/apache/parquet/scrooge/ScroogeReadSupport.java]: My bad, this is now removed.

**Comment by julienledem** [comment]
@ttim and @tlazaro, do you mind taking a look and letting us know if this looks ok to you? We'd want to get this merged.

**Comment by SinghAsDev** [comment]
> @ttim and @tlazaro, do you mind taking a look and letting us know if this looks ok to you? We'd want to get this merged.

Hey @ttim and @tlazaro, we have been using this at Pinterest for a few years just fine, and are wanting to have this merged in so that we can get rid of our internal fork. Also, this is config controlled, so default behavior remains unchanged. Would really appreciate if you can help reviewing this from twitter side, so that we can go ahead with merging this. I can rebase the patch right after.

**Comment by gszadovszky** [comment]
@SinghAsDev, could you please rebase this PR to have the conflicts resolved?

**Comment by SinghAsDev** [comment]
> @SinghAsDev, could you please rebase this PR to have the conflicts resolved?

Hey @gszadovszky this is now rebased, let me know if there is anything else I can do to have this merged, thanks!

**Review by tlazaro — APPROVED** [review]
Thanks for your work and for bringing our attention to it!

I don't have approving rights but I think this is fine. Timur will be able to check this sometime next week.

I have one question though, I would like to understand better. Couldn't we have done a version that works for any level of nesting of lists? I wouldn't want to see a 'need to support 4 level lists' in the future. Not grounds to block this PR, we can always revisit if needed and given there will be a config flag from now on, easy to enhance later.

**Comment by SinghAsDev** [comment]
> Thanks for your work and for bringing our attention to it!
> 
> I don't have approving rights but I think this is fine. Timur will be able to check this sometime next week.
> 
> I have one question though, I would like to understand better. Couldn't we have done a version that works for any level of nesting of lists? I wouldn't want to see a 'need to support 4 level lists' in the future. Not grounds to block this PR, we can always revisit if needed and given there will be a config flag from now on, easy to enhance later.

Thanks for the review!

I actually don't see a reason for having more level of nesting. The three level nesting was needed to support nulls at all levels of list definition. Moreover, this is how other models have it implemented as well.

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java
Comment (gszadovszky) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ParquetThriftBytesOutputFormat.java:83]: nit:
```suggestion
  /**
   * Configuration configuration, TProtocolFactory protocolFactory,
   * Class<? extends TBase<?, ?>> thriftClass, boolean buffered,
   * FieldIgnoredHandler errorHandler)} instead
   * @deprecated Use @link{ParquetThriftBytesOutputFormat(
   */
  @Deprecated
  public ParquetThriftBytesOutputFormat(TProtocolFactory protocolFactory,
```

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java
Comment (gszadovszky) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftBytesWriteSupport.java:95]: nit:
```suggestion
  /**
   * TProtocolFactory protocolFactory, Class<? extends TBase<?, ?>> thriftClass,
   * boolean buffered, FieldIgnoredHandler errorHandler)} instead
   * @deprecated Use @link{ThriftBytesWriteSupport(Configuration configuration,
   */
  @Deprecated
  public ThriftBytesWriteSupport(TProtocolFactory protocolFactory,
```

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java
Comment (gszadovszky) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:55]: I know it is only about this change but we always struggles with our documentation. We have started to document the available configs in [parquet-hadoop](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/README.md) and also in [parquet-avro](https://github.com/apache/parquet-mr/blob/master/parquet-avro/README.md). Maybe, it worth a similar documentation in parquet-thrift as well.

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java
Comment (gszadovszky) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java:81]: It seems this variable gets a value in the constructor anyway. I would suggest having it final and assign value in the constructor only once.

**Review by gszadovszky — CHANGES_REQUESTED** [review]
I've run through the changeset and found some minor issues.
I am not an expert in parquet-thrift so let's wait for @ttim's approval. I'll push this change if there are no objections.

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java
Comment (SinghAsDev) [reply parquet-thrift/src/main/java/org/apache/parquet/thrift/ParquetWriteProtocol.java:55]: Good point, I added a Reademe with info on config this diff adds. I can follow up with a separate diff to add more info on the Readme.

**Comment by SinghAsDev** [comment]
> I've run through the changeset and found some minor issues.
> I am not an expert in parquet-thrift so let's wait for @ttim's approval. I'll push this change if there are no objections.

Thanks @gszadovszky! Appreciate the review. I addressed the feedbacks. Could you give it a final pass and help with merging? I can address @ttim's feedbacks, if any, in a follow up diff.

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java
Comment (SinghAsDev) [reply parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java]: Good point, made the change.

**Review by gszadovszky — APPROVED** [review]

**Pull request merged**
